app/src/main/java/com/android/unio/model/functions/CloudFunctions.kt

```diff
@@ -0,0 +1,192 @@
+                                      Location::name.name to newEvent.location.name),
+                              Event::types.name to newEvent.types.map { it.name },
+                              Event::maxNumberOfPlaces.name to newEvent.maxNumberOfPlaces,
+                              Event::numberOfSaved.name to newEvent.numberOfSaved,
+                              Event::eventPictures.name to newEvent.eventPictures.uids),
+                      "isNewEvent" to isNewEvent,
+                      "associationUid" to associationUId))
+              .addOnSuccessListener { result ->
+                val responseData = result.data as? String
+                if (responseData != null) {
+                  onSuccess(responseData)
+                } else {
+                  onError(IllegalStateException("Unexpected response format from Cloud Function."))
+                }
+              }
+              .addOnFailureListener { error -> onError(error) }
+        },
+        onError = { error -> onError(error) })
+  } catch (e: Exception) {
+    onError(e)
+  }
+}
+
+/**
+ * Adds or edits a role by calling a Firebase Cloud Function to save the role.
+ *
+ * This function uploads role details to a Firebase Cloud Function, including role-specific
+ * information and permissions. It retrieves the current user's token ID and sends it to the Cloud
+ * Function, along with the role data.
+ *
+ * @param newRole The role object to be added or updated.
+ * @param associationUId The unique identifier of the association to which the role belongs.
+ * @param onSuccess A callback function that is called when the role is successfully added or
+ *   updated.
+ * @param onError A callback function that is called if an error occurs during the process.
+ * @param isNewRole A boolean value indicating whether the role is new (true) or being edited
+ *   (false).
+ */
+fun addEditRoleCloudFunction(
+    newRole: Role,
+    associationUId: String,
+    onSuccess: (String) -> Unit,
+    onError: (Exception) -> Unit,
+    isNewRole: Boolean
+) {
+  try {
+    giveCurrentUserTokenID(
+        onSuccess = { tokenId ->
+          Firebase.functions
+              .getHttpsCallable("saveRole")
+              .call(
+                  hashMapOf(
+                      "tokenId" to tokenId,
+                      "role" to
+                          mapOf(
+                              "displayName" to newRole.displayName,
+                              "permissions" to
+                                  newRole.permissions.getGrantedPermissions().toList().map {
+                                      permission ->
+                                    permission.stringName
+                                  },
+                              "color" to newRole.color.toInt(),
+                              "uid" to newRole.uid),
+                      "isNewRole" to isNewRole,
+                      "associationUid" to associationUId))
```

Comment:
You should really consider extracting this into its own serialization method, and avoid having hardcoded strings as they can easily lead to issues.

Reply:
Sure ! I now implements this : 
```kotlin
private fun serializeRoleData(
    tokenId: String,
    newRole: Role,
    associationUId: String,
    isNewRole: Boolean
): Map<String, Any> {
    return mapOf(
        "tokenId" to tokenId,
        "role" to mapOf(
            "displayName" to newRole.displayName,
            "permissions" to newRole.permissions.getGrantedPermissions().toList().map { permission ->
                permission.stringName
            },
            "color" to newRole.color.toInt(),
            "uid" to newRole.uid
        ),
        "isNewRole" to isNewRole,
        "associationUid" to associationUId
    )
}
```